### PR TITLE
PAG variant for AnimateDiff

### PR DESCRIPTION
**Pull request by a-r-r-o-w**
# What does this PR do?

Looking at #8710, I thought it might interesting to apply PAG to video generation pipelines, and see if there's interest in supporting this.

In addition to this, I would also like to propose the addition of `AutoPipelineForTextToVideo` since we support a few video models now, and this will continue to grow with ongoing research progress. WDYT?

<details>
<summary> Code </summary>

```py
import torch
from diffusers import AnimateDiffPipeline, MotionAdapter, DDIMScheduler
from diffusers.pipelines.pag.pipeline_pag_sd_animatediff import AnimateDiffPAGPipeline
from diffusers.utils import export_to_gif


# model_id = "runwayml/stable-diffusion-v1-5"
model_id = "SG161222/Realistic_Vision_V5.1_noVAE"
motion_adapter_id = "guoyww/animatediff-motion-adapter-v1-5-2"

prompt = "car, futuristic cityscape with neon lights, street, no human"
negative_prompt = "low quality, bad quality"
num_inference_steps = 25
guidance_scale = 6
pag_scale = 3.0

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id)
scheduler = DDIMScheduler.from_pretrained(model_id, subfolder="scheduler", beta_schedule="linear", steps_offset=1, clip_sample=False)
pipe = AnimateDiffPAGPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    scheduler=scheduler,
    pag_applied_layers=[],
    torch_dtype=torch.float16,
).to("cuda")

configs = [
    dict(pag_scale=0.0, clip_skip=None, free_init=False),
    dict(pag_scale=0.0, clip_skip=2, free_init=False),
    dict(pag_scale=3.0, clip_skip=None, free_init=False),
    dict(pag_scale=3.0, clip_skip=2, free_init=False),
    dict(pag_scale=3.0, clip_skip=None, free_init=True),
    dict(pag_scale=3.0, clip_skip=2, free_init=True),
    dict(pag_scale=0.5, clip_skip=None, free_init=False),
    dict(pag_scale=0.5, clip_skip=2, free_init=False),
    dict(pag_scale=0.5, clip_skip=None, free_init=True),
    dict(pag_scale=0.5, clip_skip=2, free_init=True),
    dict(pag_scale=5.0, clip_skip=None, free_init=False),
    dict(pag_scale=5.0, clip_skip=2, free_init=False),
    dict(pag_scale=5.0, clip_skip=None, free_init=True),
    dict(pag_scale=5.0, clip_skip=2, free_init=True),
]

for config in configs:
    free_init = config.pop("free_init", False)
    if free_init:
        pipe.enable_free_init(method="butterworth", use_fast_sampling=True)
    
    video = pipe(
        prompt=prompt,
        negative_prompt=negative_prompt,
        height=512,
        width=512,
        num_frames=16,
        guidance_scale=guidance_scale,
        num_inference_steps=num_inference_steps,
        generator=torch.Generator().manual_seed(42),
        **config,
    ).frames[0]

    if free_init:
        pipe.disable_free_init()

    export_to_gif(video, f"animatediff_pag-{config['pag_scale']}_clipskip-{config['clip_skip']}_freeinit-{config['free_init']}.gif")

motion_adapter = MotionAdapter.from_pretrained(motion_adapter_id)
scheduler = DDIMScheduler.from_pretrained(model_id, subfolder="scheduler", beta_schedule="linear", steps_offset=1, clip_sample=False)
pipe = AnimateDiffPAGPipeline.from_pretrained(
    model_id,
    motion_adapter=motion_adapter,
    scheduler=scheduler,
    pag_applied_layers=["mid"],
    torch_dtype=torch.float16,
).to("cuda")

for config in configs:
    free_init = config.pop("free_init", False)
    if free_init:
        pipe.enable_free_init(method="butterworth", use_fast_sampling=True)
    
    video = pipe(
        prompt=prompt,
        negative_prompt=negative_prompt,
        height=512,
        width=512,
        num_frames=16,
        guidance_scale=guidance_scale,
        num_inference_steps=num_inference_steps,
        generator=torch.Generator().manual_seed(42),
        **config,
    ).frames[0]

    if free_init:
        pipe.disable_free_init()

    export_to_gif(video, f"animatediff_pag-mid-{config['pag_scale']}_clipskip-{config['clip_skip']}_freeinit-{config['free_init']}.gif")
```
</details>

<table>
<tbody>
<tr>
<td align=center><b>pag 0, clip_skip None, free_init False</b></td>
<td align=center><b>pag 0, clip_skip 2, free_init False</b></td>
</tr>
<tr>
<td align=center></td>
<td align=center></td>
</tr>

<tr>
<td align=center><b>pag 3, clip_skip None, free_init False</b></td>
<td align=center><b>pag 3, clip_skip 2, free_init False</b></td>
</tr>
<tr>
<td align=center></td>
<td align=center></td>
</tr>

<tr>
<td align=center><b>pag 3, clip_skip None, free_init True</b></td>
<td align=center><b>pag 3, clip_skip 2, free_init True</b></td>
</tr>
<tr>
<td align=center></td>
<td align=center></td>
</tr>
</tbody>
</table>

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@yiyixuxu @asomoza @DN6

### CODE REVIEW

**Comment by asomoza** [comment]
Nice! I'll do some tests with it but looks good.

**Review by DN6 — COMMENTED** [review]
It's looking good to me. Could we add some tests and I think there are some issues with the copied from statement. I think you'd just need to run  `make fix-copies`

File: src/diffusers/models/attention_processor.py
Comment (yiyixuxu) [inline src/diffusers/models/attention_processor.py:2584]:
```suggestion
```

**Review by yiyixuxu — APPROVED** [review]
thanks! super cool!

File: src/diffusers/models/unets/unet_3d_blocks.py
Comment (a-r-r-o-w) [inline src/diffusers/models/unets/unet_3d_blocks.py:1577]: Context for why we need this change: https://github.com/huggingface/diffusers/pull/7707#discussion_r1570043527. It is the correct thing to do here. cc @DN6

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pag_utils.py:102]: Need to support motion modules self attention layers as well in PAGMixin. cc @yiyixuxu

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (a-r-r-o-w) [inline src/diffusers/pipelines/pag/pag_utils.py:88]: Did a little bit of a refactor as well to make all functions look similar-ish to `get_attn_index`. Can open a separate PR if this is out of scope. cc yiyixuxu

File: tests/pipelines/pag/test_pag_animatediff.py
Comment (a-r-r-o-w) [inline tests/pipelines/pag/test_pag_animatediff.py:135]: Sorry :(

File: tests/pipelines/pag/test_pag_animatediff.py
Comment (a-r-r-o-w) [inline tests/pipelines/pag/test_pag_animatediff.py:69]: @DN6 Might be of interest to you for AnimateDiff :eyes: Makes the tests much faster!

For [test_animatediff.py](https://github.com/huggingface/diffusers/blob/main/tests/pipelines/animatediff/test_animatediff.py):
```
real    2m38,020s
user    17m37,497s
sys     1m59,702s
```

For PAG AnimateDiff with new dummy component sizes:
```
real    1m36,838s
user    1m43,263s
sys     0m29,261s
```

**Comment by a-r-r-o-w** [comment]
After giving some more thought and observing the outputs, I feel the perturbed path for motion model seems to have either a lesser impact or negative impact on the generations. PAG works great for spatial self attention layers, that of text to image models, as we've seen from many other pipelines, however I think more experiments are needed when dealing with temporal layers like the attn in motion model. Will report back with some experiments soon.

**Comment by a-r-r-o-w** [comment]
Btw if we're okay with merging this for the generation improvements with just the spatial attn PAG processors, we can go ahead with that. Can continue experimenting with the motion models separately and revert the pag_utils changes here.

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [reply src/diffusers/pipelines/pag/pag_utils.py:88]: ok to include here!

File: src/diffusers/pipelines/pag/pag_utils.py
Comment (yiyixuxu) [reply src/diffusers/pipelines/pag/pag_utils.py:102]: nice, let's add a note here too https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pag/pag_utils.py#L129

File: tests/pipelines/pag/test_pag_animatediff.py
Comment (yiyixuxu) [reply tests/pipelines/pag/test_pag_animatediff.py:135]: it is probably because we didn't handle the deprecated unet config in `__init__`
related here: https://github.com/huggingface/diffusers/pull/7564

because this pipeline shares the same checkpoints with sd1.5, technically you need to handle the deprecation too even though it is a new pipeline https://github.com/huggingface/diffusers/blob/3b37fefee99425286984a9d5fa4f1850064d01eb/src/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L192

however in practise, I think maybe no one will use these really old checkpoints on animate diff, so I'm ok to skip the test here

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Comment by yiyixuxu** [comment]
>Btw if we're okay with merging this for the generation improvements with just the spatial attn PAG processors, we can go ahead with that. Can continue experimenting with the motion models separately and revert the pag_utils changes here.

what do you mean here? do you mean applying PAG only on the motion modules does not generate good results? e.g. `pag_applied_layers = ["down.block_1.motion_modules_0"]`, if you do something like "mid", it will apply on both temporal and spatial, no? would the results be same as spatial only? can we see some results

**Comment by a-r-r-o-w** [comment]
> what do you mean here? do you mean applying PAG only on the motion modules does not generate good results? e.g. `pag_applied_layers = ["down.block_1.motion_modules_0"]`, if you do something like "mid", it will apply on both temporal and spatial, no? would the results be same as spatial only? can we see some results

Thanks for the review! So, I added the changes for PAG to work with motion modules self attn (temporal layers) later [in this commit](https://github.com/huggingface/diffusers/pull/8789/commits/8265d099e95b3f4768aa43513624b3db60c7f063)). The results posted in the PR description utilize PAG in only the spatial layers. After my latest changes, I prefer the outputs of just spatial as opposed to spatial and temporal PAG from the few quick experiments I tried. I will soon look into it more thoroughly and report back. cc @asomoza

**Comment by a-r-r-o-w** [comment]
I took a quick look at the Comfy implementation earlier today to see what Kosinkadink was using: https://github.com/Kosinkadink/ComfyUI-AnimateDiff-Evolved/blob/4dd592e9fce9ac59edadee40cf4d2069165dc226/animatediff/cfg_extras.py#L59

It is being applied to both spatial and temporal attn1 layers, which is what we have at the moment too, so I'm okay to roll with this and prepare for merge. But, I think we can investigate a bit further eventually into the dynamics of PAG with temporal layers especially since there are a few unofficial implementations looking into this phenomenon making use of the latest diffusers PAGMixin implementation (example: https://github.com/pixeli99/Spatio-Temporal-Shuffle-Guidance)

**Comment by yiyixuxu** [comment]
>The results posted in the PR description utilize PAG in only the spatial layers

I think in the PR description code, you used `pag_applied_layers="mid"` -  that means it's applying to both spatial and temporal layers, no? 

the commit you added here [in this commit](https://github.com/huggingface/diffusers/pull/8789/commits/8265d099e95b3f4768aa43513624b3db60c7f063) allows you to apply PAG to ONLY temporal layers, that does not yield good results - do I understand this correctly?

**Comment by a-r-r-o-w** [comment]
> I think in the PR description code, you used `pag_applied_layers="mid"` - that means it's applying to both spatial and temporal layers, no?

Correct. With the latest commits, using `pag_applied_layers="mid"` will apply it to both spatial and temporal layers.

What I'm trying to point out is that the demos that you see in the description are the ones with only spatial PAG because I had not implemented it for motion models yet then.

> the commit you added here [in this commit](https://github.com/huggingface/diffusers/pull/8789/commits/8265d099e95b3f4768aa43513624b3db60c7f063) allows you to apply PAG to ONLY temporal layers, that does not yield good results - do I understand this correctly?

Nope. Because of that commit, PAG applies to BOTH spatial and temporal layers. I preferred the output of what was before that commit (that is ONLY spatial). Hope it makes sense now :sweat_smile:

**Comment by yiyixuxu** [comment]
you can see that here https://github.com/huggingface/diffusers/blob/3b37fefee99425286984a9d5fa4f1850064d01eb/src/diffusers/pipelines/pag/pag_utils.py#L112

when pass `pag_applied_layers="mid"`, it apply PAG to all self-attentions with name starting with "mid", that will include both spatial and temporal, with or without the commit

**Comment by yiyixuxu** [comment]
ohh i see now, we needed the change applied to `get_block_type`

**Comment by a-r-r-o-w** [comment]
maybe that is correct, but when I was debugging, I did not notice any of the motion module self attn layers being set, and that was one reason for adding my changes. The other reason being more fine grained control to do things like `mid_block_0.motion_modules...`. Let me try reverting my changes and verifying why they were needed first thing I'm awake next.

**Comment by yiyixuxu** [comment]
ok, we can merge this PR as it is consistent with comfy

**Comment by a-r-r-o-w** [comment]
Looking good to merge from my end after we merge #8846 since [this change](https://github.com/huggingface/diffusers/pull/8789#discussion_r1674793584) is out of scope of the PR.

**Comment by a-r-r-o-w** [comment]
Fixed the broken tests here. This looks good to merge after #8846 which is now complete too

**Review by DN6 — APPROVED** [review]

**Pull request merged**
